### DISCUSSION TITLE
Generated Types & Response Structures

### DISCUSSION THREAD

**Discussion by jchirk** [odata2ts/odata2ts#220] [Q&A] [answered]
I saw the config parameter `v2ModelsWithExtraResultsWrapping` allowing **collection properties** to have the `results` object wrapper.
But when directly consuming entitySets (e.g. [Northwind Customers](https://services.odata.org/V2/Northwind/Northwind.svc/Customers?$format=json)), we have also a `results` object :
```
{
  "d" : {
    "results": [  ...  ]
}
```

The idea would be to also generate the entitySet interface :
```
export interface Customer { ... }
export interface Customers { 
  results: Array<Customer>
}
```

Is it relevant or am I missing something ?

**Comment by texttechne** [comment]
Hi @jchirk,
please read [the documentation for this configuration](https://odata2ts.github.io/docs/generator/configuration#v2-extra-results-wrapper). It basically says that it only applies to **expanded collections**, because this part is not well covered in the OData V2 spec....

The examples you mention are well specified in OData V2 and covered by the `odata-core` package: [See here](https://github.com/odata2ts/odata2ts/blob/main/packages/odata-core/src/ResponseModelV2.ts). So you could do:
```
import { ODataCollectionResponseV2 }  from "@odata2ts/odata-core";

export type Customers = ODataCollectionResponseV2<Customer>
```

If you think that the documentation could be improved in this regard, I'm all ears...

**Comment by jchirk** [comment]
Hi @texttechne,

Thanks for your answer. I guess it would be useful to add your remark on that (and the solution you proposed).

Just so I understand correctly, in the OData V2 doc, it seems pretty clear that there is a ["results" property](https://www.odata.org/documentation/odata-version-2-0/json-format/#:~:text=6.%20Representing%20Collections%20of%20Entries) for the JSON format. Is it because of the Atom format that you chose not to add it ?

**Comment by texttechne** [comment, marked as answer]
Hi @jchirk,

> Just so I understand correctly, in the OData V2 doc, it seems pretty clear that there is a ["results" property](https://www.odata.org/documentation/odata-version-2-0/json-format/#:~:text=6.%20Representing%20Collections%20of%20Entries) for the JSON format. Is it because of the Atom format that you chose not to add it ?

No, Atom has nothing to do with that. It's the architecture of OData in itself actually. If you take a look at the metadata (EDMX) then you will find `EntityType` elements which in the end will correspond to a table in your DB (roughly). These EntityTypes can be used in different places: 
* in the form of an `EntitySet` they serve as entry point into the ODataService
* an entity can also reference another entity via `NavigationProperty` elements
  * and you can make up a 1:1 relationship
  * as well as a 1:n or n:m relationship (you will find `Collection(namespace.MyEntityType)`)

This all means, that the EntityType is the non-moving part for which we need an interface in TS. It can then be used in different places and will be contained in a structure dependend on the actual use case. So for you specific question: You're querying for an entity set, so your response structure will be (in V2): 
```
{
  d: {
    results: Array<MyEntityType>
  }
}
```
If you query for a single entity:
```
{
  d: MyEntityType
}
```
So it's more or less about the differentiation between moving and non-moving parts: The entity definition is fixed (non-moving part), it is used in different structures though (the moving parts which depend on your use case).

**Reply by jchirk** [reply]
Hi @texttechne,

Oh yeah, I forgot when you query a single entityType. Makes totally sense now, and your 'workaround' with `ODataCollectionResponseV2` is a neat way to deal with this issue.
Thank you for your time and for your answer ! 👍

**Comment by texttechne** [comment]
Hi @jchirk,

changed this from an issue into a discussion as it feels more appropriate.
